rvgo/test/syscall_test.go

```diff
@@ -0,0 +1,419 @@
+package test
+
+import (
+	"fmt"
+	"os"
+	"testing"
+
+	preimage "github.com/ethereum-optimism/optimism/op-preimage"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/stretchr/testify/require"
+
+	"github.com/ethereum-optimism/asterisc/rvgo/fast"
+	"github.com/ethereum-optimism/asterisc/rvgo/slow"
+)
+
+var syscallInsn = []byte{0x73}
+
+func staticOracle(t *testing.T, preimageData []byte) *testOracle {
+	return &testOracle{
+		hint: func(v []byte) {},
+		getPreimage: func(k [32]byte) []byte {
+			if k != preimage.Keccak256Key(crypto.Keccak256Hash(preimageData)).PreimageKey() {
+				t.Fatalf("invalid preimage request for %x", k)
+			}
+			return preimageData
+		},
+	}
+}
+
+func runEVM(t *testing.T, contracts *Contracts, addrs *Addresses, stepWitness *fast.StepWitness, fastPost fast.StateWitness) {
+	env := newEVMEnv(t, contracts, addrs)
+	evmPost, _, _ := stepEVM(t, env, stepWitness, addrs, 0)
+	require.Equal(t, hexutil.Bytes(fastPost).String(), hexutil.Bytes(evmPost).String(),
+		"fast VM produced different state than EVM")
+}
+
+func runSlow(t *testing.T, stepWitness *fast.StepWitness, fastPost fast.StateWitness, po slow.PreimageOracle) {
+	slowPostHash, err := slow.Step(stepWitness.EncodeStepInput(fast.LocalContext{}), po)
+	require.NoError(t, err)
+	fastPostHash, err := fastPost.StateHash()
+	require.NoError(t, err)
+	require.Equal(t, fastPostHash, slowPostHash, "fast VM produced different state than slow VM")
+}
+
+func TestStateSyscallUnsupported(t *testing.T) {
+	syscalls := []int{
+		fast.SysPrlimit64,
+		fast.SysFutex,
+		fast.SysNanosleep,
+	}
+
+	for _, syscall := range syscalls {
+		t.Run(fmt.Sprintf("sys_%d", syscall), func(t *testing.T) {
+			pc := uint64(0)
+			state := &fast.VMState{
+				PC:              pc,
+				Heap:            0,
+				ExitCode:        0,
+				Exited:          false,
+				Memory:          fast.NewMemory(),
+				LoadReservation: 0,
+				Registers:       [32]uint64{17: uint64(syscall)},
+				Step:            0,
+			}
+			state.Memory.SetUnaligned(pc, syscallInsn)
+
+			fastState := fast.NewInstrumentedState(state, nil, os.Stdout, os.Stderr)
+			_, err := fastState.Step(true)
+			var syscallErr *fast.UnsupportedSyscallErr
+			require.ErrorAs(t, err, &syscallErr)
+
+			// TODO: Test EVM & slow VM
```

Comment:
Can you please add the reason why we cannot test slow/EVM mode here?

Reply:
Currently most codes for testing EVM and slow VM are not designed to test steps that must fail. To do that, we need to modify them to ignore errors optionally. it seems to be a large diffs than I expected. so I'll do it as another task.